### PR TITLE
Add worker num as one of load resource

### PR DESCRIPTION
**Pull request by yah01**
/kind improvement
related #23966

### CODE REVIEW

File: internal/querynodev2/segments/segment_loader.go [hunk 1]
```diff
@@ -287,53 +296,65 @@ func (loader *segmentLoader) notifyLoadFinish(segments ...*querypb.SegmentLoadIn
...
+	diskUsage, err := GetLocalUsedSize(paramtable.Get().LocalStorageCfg.Path.GetValue())
+	if err != nil {
+		return resource, 0, errors.Wrap(err, "get local used size failed")
+	}
+	diskCap := paramtable.Get().QueryNodeCfg.DiskCapacityLimit.GetAsUint64()
+
 	loader.mut.Lock()
 	defer loader.mut.Unlock()
 
+	if loader.committedResource.WorkNum >= runtime.NumCPU()*paramtable.Get().CommonCfg.ThreadCoreCoefficient.GetAsInt() ||
```
Comment (congqixia) [inline internal/querynodev2/segments/segment_loader.go:314]: Why work num insufficient shall fail here?

File: internal/querynodev2/segments/segment_loader.go [hunk 2]
```diff
@@ -287,53 +296,65 @@ func (loader *segmentLoader) notifyLoadFinish(segments ...*querypb.SegmentLoadIn
...
+	}
+
 	concurrencyLevel := funcutil.Min(runtime.GOMAXPROCS(0), len(infos))
 
-	var memUsage, diskUsage uint64
 	for _, info := range infos {
-		logNum := 0
 		for _, field := range info.GetBinlogPaths() {
-			logNum += len(field.GetBinlogs())
+			resource.WorkNum += len(field.GetBinlogs())
```
Comment (congqixia) [inline internal/querynodev2/segments/segment_loader.go:324]: Scalar field does not load with C++ chunk manager

File: internal/querynodev2/segments/segment_loader.go [hunk 3]
```diff
@@ -287,53 +296,65 @@ func (loader *segmentLoader) notifyLoadFinish(segments ...*querypb.SegmentLoadIn
...
+	}
+	diskCap := paramtable.Get().QueryNodeCfg.DiskCapacityLimit.GetAsUint64()
+
 	loader.mut.Lock()
 	defer loader.mut.Unlock()
 
+	if loader.committedResource.WorkNum >= runtime.NumCPU()*paramtable.Get().CommonCfg.ThreadCoreCoefficient.GetAsInt() ||
+		loader.committedResource.MemorySize+memoryUsage >= totalMemory ||
+		loader.committedResource.DiskSize+uint64(diskUsage) >= diskCap {
+		return resource, 0, merr.WrapErrServiceMemoryLimitExceeded(float32(loader.committedResource.MemorySize+memoryUsage), float32(totalMemory))
```
Comment (congqixia) [inline internal/querynodev2/segments/segment_loader.go:317]: When request fails with `WorkNum` insufficient, returning `ErrServiceMemoryLimitExceeded` may cause confusion

**Review by congqixia — APPROVED** [review]
/lgtm

File: pkg/util/paramtable/param_item.go [hunk 4]
```diff
@@ -193,6 +197,12 @@ func getAsInt64(v string) int64 {
 	}, 0)
 }
 
+func getAsUnt64(v string) uint64 {
```
Comment (jiaoew1991) [inline pkg/util/paramtable/param_item.go:200]: getAsUint64

File: internal/querynodev2/segments/segment_loader.go [hunk 5]
```diff
@@ -75,6 +71,24 @@ type Loader interface {
 	LoadIndex(ctx context.Context, segment *LocalSegment, info *querypb.SegmentLoadInfo) error
 }
 
+type LoadResource struct {
+	MemorySize uint64
+	DiskSize   uint64
+	WorkNum    int
+}
+
+func (r *LoadResource) Add(resource LoadResource) {
```
Comment (jiaoew1991) [inline internal/querynodev2/segments/segment_loader.go:80]: LoadResource or *LoadResource 🤔？

Comment (yah01) [reply internal/querynodev2/segments/segment_loader.go:80]: Here it need to modify the struct, so receiver is a pointer. The struct is simple so we pass it by value to optimize for GC

**Review by congqixia — APPROVED** [review]
/lgtm

**Pull request merged**